iroh-bytes/src/downloader.rs

```diff
@@ -454,540 +506,535 @@ struct Service<G: Getter, D: Dialer> {
-            let kind = kind.into_inner();
-            Some((kind, info))
-        } else {
-            None
+            self.remove_kind_from_provider_map(&kind);
         }
     }
 
     /// Handle receiving a new connection.
     fn on_connection_ready(&mut self, node: NodeId, result: anyhow::Result<D::Connection>) {
         match result {
             Ok(connection) => {
-                trace!(%node, "connected to node");
+                trace!(node=%node.fmt_short(), "connected to node");
                 let drop_key = self.goodbye_nodes_queue.insert(node, IDLE_PEER_TIMEOUT);
                 self.nodes
                     .insert(node, ConnectionInfo::new_idle(connection, drop_key));
-                self.on_node_ready(node);
             }
             Err(err) => {
                 debug!(%node, %err, "connection to node failed")
             }
         }
     }
 
-    /// Called after the connection to a node is established, and after finishing a download.
-    ///
-    /// Starts the next provider hash download, if there is one.
-    fn on_node_ready(&mut self, node: NodeId) {
-        // Get the next provider hash for this node.
-        let Some(hash) = self.providers.get_next_provider_hash_for_node(&node) else {
-            return;
-        };
-
-        if self.is_current_request(hash) {
-            return;
-        }
-
-        let Some(conn) = self.get_node_connection_for_download(&node) else {
-            return;
-        };
-
-        let Some((kind, info)) = self.unschedule(hash) else {
-            debug_assert!(
-                false,
-                "invalid state: expected {hash:?} to be scheduled, but it wasn't"
-            );
-            return;
-        };
-
-        let PendingRequestInfo {
-            intents,
-            remaining_retries,
-            ..
-        } = info;
-
-        self.start_download(kind, node, conn, remaining_retries, intents);
-    }
-
-    fn on_download_completed(&mut self, kind: DownloadKind, result: Result<(), FailureAction>) {
+    async fn on_download_completed(&mut self, kind: DownloadKind, result: InternalDownloadResult) {
         // first remove the request
-        let info = self
-            .current_requests
+        let active_request_info = self
+            .active_requests
             .remove(&kind)
             .expect("request was active");
 
-        // update the active requests for this node
-        let ActiveRequestInfo {
-            intents,
-            node,
-            mut remaining_retries,
-            ..
-        } = info;
+        // get general request info
+        let request_info = self.requests.remove(&kind).expect("request was active");
 
+        let ActiveRequestInfo { node, temp_tag, .. } = active_request_info;
+
+        // get node info
         let node_info = self
             .nodes
             .get_mut(&node)
             .expect("node exists in the mapping");
```

Comment:
@ppodolsky triggered this `expect` when using this branch. I have a suspicion and am writing a test to confirm.

Reply:
This is now fixed.